zip-0315.rst

```diff
@@ -1,9 +1,629 @@
+Sapling funds, do so.
+
+The user's consent is needed to reveal amounts publically (as opposed
+to revealing them to the holder of a viewing key authorized to see that
+amount). Therefore, there should be per-transaction opt-in for any
+transfer that publically reveals amounts on chain.
+
+* there may be a compatibility issue for amount-revealing cross-pool txns that were
+  previously allowed without opt-in
+
+Wallets MUST NOT automatically combine funds across pools to satisfy a transfer
+(since that could reveal the total funds the user holds in some pool).
+
+In order to maintain the integrity of IVK guarantees, wallets should not generate
+unified addresses that contain internal receivers, nor expose internal receivers
+(such as those used for auto-shielding and change outputs) in any way.
+
+Open questions:
+
+* should there be an auto-migration option from Sapling to Orchard?
+
+# str4d notes
+
+If we want to have both automatic and opportunistic shielding, and keep the two
+indistinguishable, then we can't auto-shield when the transparent balance reaches
+some threshold (otherwise opportunistic would either never be used, or would be
+identifiable when it uses the balance below the threshold).
+
+Instead, a proposition: we define a distribution of "time since last payment to the
+address" from which we sample the time at which the auto-shielding transaction will
+be created. This distribution is weighted by the balance in the address, so as more
+funds accrue, the auto-shielding transaction is more likely to be created.
+
+- It ensures that all funds will eventually be auto-shielded, while preventing
+  fee-dusting attacks (where dust is sent in order to repeatedly consume fees from
+  the wallet), as the auto-shielding transaction is not directly triggered by payment
+  receipt.
+
+- If the user makes a shielding transaction in the meantime, we opportunistically
+  shield, without it being clearly not an auto-shielding transaction.
+
+- If a wallet is offline for a long time, then it would likely auto-shield as soon as
+  it finishes syncing. This maybe isn't enough to reveal that the wallet came online,
+  except that it _might_ result in auto-shielding transactions for multiple
+  transparent addresses being created at the same time. So we might want to
+  special-case this?
+
+Properties we want from auto-shielding:
+
+- Auto-shielding transactions MUST NOT shield from multiple transparent receivers in
+  the same transaction.
+  - Doing so would trivially link diversified UAs containing transparent receivers.
+
+Properties we want from auto-migration:
+
+- Receipt of a shielded payment MUST NOT trigger any on-chain behaviour (as that
+  reveals transaction linkability).
+
+Both auto-shielding and auto-migration are time-triggered actions, not
+receipt-triggered actions. An auto-shielding or auto-migration transaction MUST NOT
+be created as a direct result of a payment being received.
```

Comment:
Not directly related to this paragraph, but a thought: two wallet could theoretically enter a ZEC-destructive war. If one wallet does auto-migration to Orchard while another other wallet on the same account has a goal to maintain a balance between sapling and orchard, these two wallets may continually auto-spend across pools till fees deplete the account.
For this reason, IMO _automatically_ moving funds across shielded pools (either direction) is a high risk thing. Either do it as part of another transaction or only by explicit user gesture.
I think this should be called out to prevent two wallet apps from fighting.

Reply:
I don't understand this comment. Do you mean two wallet apps sharing the same seed? That's something that a user should NEVER do.

Along those lines, I think that we should have an explicit section in this ZIP intended for wallet users, in addition to all the information that's intended for wallet developers. Now, this probably can't be expressed as explicit directives, but what we should do instead is define a number of user profiles with different threat models, and describe the threat model that they should consider when using a wallet that follows the ZIP 315 directives.

Reply:
Yes, I'm talking about two wallets with the same seed. I've never heard the suggestion that I shouldn't sync two wallets with the same seed. I can't imagine why. When you say two wallet _apps_ are you specifically saying a user should not trust _two vendors_ of apps with the same seed, or are you saying I shouldn't even have one vendor's app installed on two devices with the same seed?

I don't see the danger either way. A compromise of either device or a rogue app can result in loss of your ZEC. Two devices or two apps, double the risk. But in such a case, the ZEC will be stolen whether the seed phrase also appears on another device or not.
And I for one have found it useful many times to have access to my Zcash from multiple devices.